features/net/network-socket/WiFiInterface.h

```diff
@@ -43,28 +40,78 @@ class WiFiInterface: public NetworkInterface
+     *  you'll need to query @a get_state until it's in NSAPI_IF_STATE_CONNECTED state, otherwise the @a cb will be
+     *  called with connection results.
+     *
+     *  Note: @a ssid and @a pass must be kept until the connection is made, that is the callback has been called or the
+     *        state changed to @a NSAPI_IF_STATE_CONNECTED as they are passed by value.
      *
      *  @param ssid      Name of the network to connect to
      *  @param pass      Security passphrase to connect to the network
      *  @param security  Type of encryption for connection
-     *  @return          0 on success, negative error code on failure
+     *  @param cb        Function to be called when the connect finishes
+     *  @param data      Arbitrary user data to pass to @a cb function
      */
-    virtual int connect(const char *ssid, const char *pass, nsapi_security_t security = NSAPI_SECURITY_NONE) = 0;
+    virtual void connect_async(const char *ssid, const char *pass,nsapi_security_t security = NSAPI_SECURITY_NONE,
+                               wifi_connect_cb_t cb = NULL, void *data = NULL) = 0;
 
     /** Stop the interface
      *
-     *  @return          0 on success, negative error code on failure
+     *  @return          0 on success, or error code on failure
      */
-    virtual int disconnect() = 0;
+    virtual nsapi_error_t disconnect() = 0;
 
     /** Get the local MAC address
      *
      *  @return         Null-terminated representation of the local MAC address
      */
     virtual const char *get_mac_address() = 0;
-};
 
+    /** Get the current WiFi interface state
+     *
+     * @returns         Interface state enum
+     */
+    virtual nsapi_if_state_t get_state() = 0;
+
+    /** Gets the current radio signal strength for active connection
+     *
+     * @return Connection strength in dBm (negative value), or 0 if measurement impossible
+     */
+    virtual int8_t get_rssi() = 0;
+
+    /** Scan for available networks
+     *
+     * This function will block.
+     *
+     * @param  ap    Pointer to allocated array to store discovered AP
+     * @param  count Size of allocated @a res array, or 0 to only count available AP
+     * @return       Number of entries in @a, or if @a count was 0 number of available networks, negative on error
+     *               see @a nsapi_error
+     */
+    virtual int scan(wifi_ap_t *res, unsigned count) = 0;
+
+    /** Scan for available networks
+     *
+     * This function won't block, it'll return immediately and call provided callback for each discovered network with
+     * AP data and user @a data. After the last discovered network @a cb will be called with NULL as @a ap, so it
+     * will be always called at least once.
+     *
+     * @param  cb    Function to be called for every discovered network
+     * @param  data  A user handle that will be passed to @a cb along with the AP data
+     */
+    virtual void scan(wifi_ap_scan_cb_t cb, void *data = NULL) = 0;
```

Comment:
Nitpick (sorry, it's Friday): I'd call this something like `scan_async` (and maybe even `scan_sync` for the blocking version). Even though the documentation is very clear, calling both functions `sync` and having one be synchronous and the other one asynchronous can lead to some confusion for people reading the code. Makes even more sense since you already did this with `connect_async`.